File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Services/ServiceProcess.vb
' 09-13-06

Imports System.Text
Imports System.Threading
Imports Tva.DateTime.Common

Public Class ServiceProcess

    Private m_processThread As Thread
    Private m_name As String
    Private m_parameters As Object()
    Private m_executionMethod As ExecutionMethodSignature
    Private m_serviceHelper As ServiceHelper
    Private m_currentState As ProcessState
    Private m_lastExecutionTime As Double

    Public Delegate Sub ExecutionMethodSignature(ByVal name As String, ByVal parameters As Object())

    Public Sub New(ByVal executionMethod As ExecutionMethodSignature, ByVal name As String, _
            ByVal serviceHelper As ServiceHelper)
        MyClass.New(executionMethod, name, Nothing, serviceHelper)
    End Sub

    Public Sub New(ByVal executionMethod As ExecutionMethodSignature, ByVal name As String, _
            ByVal parameters As Object(), ByVal serviceHelper As ServiceHelper)
        m_name = name
        m_parameters = parameters
        m_executionMethod = executionMethod
        m_serviceHelper = serviceHelper
        m_currentState = ProcessState.Unprocessed
    End Sub

    Public Property Name() As String
        Get
            Return m_name
        End Get
        Set(ByVal value As String)
            If Not String.IsNullOrEmpty(value) Then
                m_name = value
            Else
                Throw New ArgumentException("Name cannot be null.")
            End If
        End Set
    End Property

    Public Property Parameters() As Object()
        Get
            Return m_parameters
        End Get
        Set(ByVal value As Object())
            m_parameters = value
        End Set
    End Property

    Public Property ExecutionMethod() As ExecutionMethodSignature
        Get
            Return m_executionMethod
        End Get
        Set(ByVal value As ExecutionMethodSignature)
            m_executionMethod = value
        End Set
    End Property

    Public Property ServiceHelper() As ServiceHelper
        Get
            Return m_serviceHelper
        End Get
        Set(ByVal value As ServiceHelper)
            m_serviceHelper = value
        End Set
    End Property

    Public Property CurrentState() As ProcessState
        Get
            Return m_currentState
        End Get
        Private Set(ByVal value As ProcessState)
            m_currentState = value
            m_serviceHelper.ProcessStateChanged(m_name, m_currentState)
        End Set
    End Property

    Public ReadOnly Property LastExecutionTime() As Double
        Get
            Return m_lastExecutionTime
        End Get
    End Property

    Public ReadOnly Property Status() As String
        Get
            With New StringBuilder()
                .Append("              Process Name: ")
                .Append(m_name)
                .Append(Environment.NewLine)
                .Append("             Current State: ")
                .Append(m_currentState.ToString())
                .Append(Environment.NewLine)
                .Append("       Last Execution Time: ")
                .Append(SecondsToText(m_lastExecutionTime))
                .Append(Environment.NewLine)

                Return .ToString()
            End With
        End Get
    End Property

    Public Sub Start()

        m_processThread = New Thread(AddressOf InvokeExecutionMethod)
        m_processThread.Start()

    End Sub

    Public Sub Abort()

        If m_processThread IsNot Nothing Then m_processThread.Abort()
        Me.CurrentState = ProcessState.Aborted

    End Sub

    Private Sub InvokeExecutionMethod()

        If m_executionMethod IsNot Nothing Then
            CurrentState = ProcessState.Processing
            Dim startTime As Long = Date.Now.Ticks
            m_executionMethod.Invoke(m_name, m_parameters)
            m_lastExecutionTime = TicksToSeconds(Date.Now.Ticks - startTime)
            CurrentState = ProcessState.Processed
        End If
        m_processThread = Nothing

    End Sub

End Class

File: Source/OldCode/Archive/2005 TVA Code Library/Version/TVACodeLibrary/TVA.Services_CS/ServiceProcess.cs
using System.Diagnostics;
using System.Linq;
using System.Collections;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.Text;
using System.Threading;
//using TVA.DateTime.Common;
using TVA.Threading;

// 09-13-06


namespace TVA.Services
{
	public class ServiceProcess
	{
		
		
		#if ThreadTracking
		private ManagedThread m_processThread;
		#else
		private Thread m_processThread;
		#endif
		private string m_name;
		private object[] m_parameters;
		private ExecutionMethodSignature m_executionMethod;
		private ServiceHelper m_serviceHelper;
		private ProcessState m_currentState;
		private DateTime m_executionStartTime;
		private DateTime m_executionStopTime;
		
		public delegate void ExecutionMethodSignature(string name, object[] parameters);
		
		public ServiceProcess(ExecutionMethodSignature executionMethod, string name, ServiceHelper serviceHelper) : this(executionMethod, name, null, serviceHelper)
		{
		}
		
		public ServiceProcess(ExecutionMethodSignature executionMethod, string name, object[] parameters, ServiceHelper serviceHelper)
		{
			m_name = name;
			m_parameters = parameters;
			m_executionMethod = executionMethod;
			m_serviceHelper = serviceHelper;
			m_currentState = processState.Unprocessed;
		}
		
		public string Name
		{
			get
			{
				return m_name;
			}
			set
			{
				if (! string.IsNullOrEmpty(value))
				{
					m_name = value;
				}
				else
				{
					throw (new ArgumentException("Name cannot be null."));
				}
			}
		}
		
		public object[] Parameters
		{
			get
			{
				return m_parameters;
			}
			set
			{
				m_parameters = value;
			}
		}
		
		public ExecutionMethodSignature ExecutionMethod
		{
			get
			{
				return m_executionMethod;
			}
			set
			{
				m_executionMethod = value;
			}
		}
		
		public ServiceHelper ServiceHelper
		{
			get
			{
				return m_serviceHelper;
			}
			set
			{
				m_serviceHelper = value;
			}
		}
		
		public ProcessState CurrentState
		{
			get
			{
				return m_currentState;
			}
			private set
			{
				m_currentState = value;
				m_serviceHelper.ProcessStateChanged(m_name, m_currentState);
			}
		}
		
		public DateTime ExecutionStartTime
		{
			get
			{
				return m_executionStartTime;
			}
		}
		
		public DateTime ExecutionStopTime
		{
			get
			{
				return m_executionStopTime;
			}
		}
		
		public double LastExecutionTime
		{
			get
			{
				return TVA.DateTime.Common.TicksToSeconds(m_executionStopTime.Ticks - m_executionStartTime.Ticks);
			}
		}
		
		public string Status
		{
			get
			{
				System.Text.StringBuilder with_1 = new StringBuilder();
				with_1.Append("              Process Name: ");
				with_1.Append(m_name);
				with_1.AppendLine();
				with_1.Append("             Current State: ");
				with_1.Append(m_currentState.ToString());
				with_1.AppendLine();
				with_1.Append("      Execution Start Time: ");
				if (m_executionStartTime != DateTime.MinValue)
				{
					with_1.Append(m_executionStartTime.ToString());
				}
				else
				{
					with_1.Append("N/A");
				}
				with_1.AppendLine();
				with_1.Append("       Execution Stop Time: ");
				if (m_executionStopTime != DateTime.MinValue)
				{
					with_1.Append(m_executionStopTime.ToString());
				}
				else
				{
					with_1.Append("N/A");
				}
				with_1.AppendLine();
				with_1.Append("       Last Execution Time: ");
				with_1.Append(TVA.DateTime.Common.SecondsToText(this.LastExecutionTime));
				with_1.AppendLine();
				
				return with_1.ToString();
			}
		}
		
		public void Start()
		{
			
			// Start the execution on a seperate thread.
			#if ThreadTracking
			m_processThread = new ManagedThread(InvokeExecutionMethod);
			m_processThread.Name = "TVA.Services.ServiceProcess.InvokeExecutionMethod() [" + m_name + "]";
			#else
			m_processThread = new Thread(new System.Threading.ThreadStart(InvokeExecutionMethod));
			#endif
			m_processThread.Start();
			
		}
		
		public void Abort()
		{
			
			if (m_processThread != null)
			{
				// We'll abort the process only if it is currently executing.
				m_processThread.Abort();
			}
			
		}
		
		private void InvokeExecutionMethod()
		{
			
			if (m_executionMethod != null)
			{
				CurrentState = processState.Processing;
				m_executionStartTime = DateTime.Now;
				m_executionStopTime = DateTime.MinValue;
				try
				{
					// We'll keep the invokation of the delegate in Try...Catch to absorb any exceptions that
					// were not handled by the consumer.
					m_executionMethod(m_name, m_parameters);
					this.CurrentState = processState.Processed;
				}
				catch (ThreadAbortException)
				{
					this.CurrentState = processState.Aborted;
				}
				catch (Exception)
				{
					// We'll absorb any exceptions if unhandled by the client.
					this.CurrentState = processState.Exception;
				}
				finally
				{
					m_executionStopTime = DateTime.Now;
				}
			}
			m_processThread = null;
			
		}
		
	}
	
}
